src/R/Components/Impl/Plots/Implementation/RPlotManager.cs

```diff
@@ -188,35 +283,143 @@ internal class RPlotManager : IRPlotManager {
+        private async Task ProcessPlotMessage(IRPlotDeviceViewModel viewModel, PlotMessage plot) {
+            if (plot.IsClearAll) {
+                await viewModel.PlotMessageClearAllAsync(plot.DeviceId, plot.DeviceNum);
+                History.RemoveAll(plot.DeviceId);
+            } else if (plot.IsPlot) {
+                try {
+                    var img = plot.ToBitmapImage();
+                    await viewModel.PlotMessageLoadPlotAsync(plot.DeviceId, plot.PlotId, img, plot.DeviceNum, plot.ActivePlotIndex, plot.PlotCount);
+                    History.AddOrUpdate(viewModel.DeviceName, viewModel.DeviceId, plot.PlotId, viewModel.SessionProcessId, img);
+                } catch (Exception e) when (!e.IsCriticalException()) {
+                    await viewModel.PlotMessageLoadErrorAsync(plot.DeviceId, plot.PlotId, plot.DeviceNum, plot.ActivePlotIndex, plot.PlotCount);
+                }
+            } else if (plot.IsError) {
+                await viewModel.PlotMessageLoadErrorAsync(plot.DeviceId, plot.PlotId, plot.DeviceNum, plot.ActivePlotIndex, plot.PlotCount);
+            }
+        }
 
-        public Task ExportToPdfAsync(string outputFilePath) =>
-            ExportAsync(outputFilePath, _interactiveWorkflow.RSession.ExportToPdfAsync(outputFilePath, PixelsToInches(_lastPixelWidth), PixelsToInches(_lastPixelHeight)));
+        private async Task RemoveFromHistoryAsync(Guid plotId) {
+            await InteractiveWorkflow.Shell.SwitchToMainThreadAsync();
+            History.Remove(plotId);
+        }
 
-        public void EndLocatorMode() {
-            EndLocatorMode(LocatorResult.CreateNotClicked());
+        private async Task ClearHistoryAsync() {
+            await InteractiveWorkflow.Shell.SwitchToMainThreadAsync();
+            History.Clear();
         }
 
-        public void EndLocatorMode(LocatorResult result) {
-            var tcs = _locatorTcs;
-            _locatorTcs = null;
-            tcs?.SetResult(result);
-            _interactiveWorkflow.Shell.DispatchOnUIThread(() => SetLocatorModeUI(false));
-            LocatorModeChanged?.Invoke(this, null);
+        private async void RSession_Mutated(object sender, EventArgs e) {
+            await RSessionMutatedAsync();
         }
 
-        private void SetLocatorModeUI(bool locatorMode) {
-            if (VisualComponent != null) {
-                VisualComponent.SetLocatorMode(locatorMode);
+        private async Task RSessionMutatedAsync() {
+            try {
+                var deviceId = await InteractiveWorkflow.RSession.GetActivePlotDeviceAsync();
+                var deviceChanged = ActiveDeviceId != deviceId;
+                ActiveDeviceId = deviceId;
+
+                // Update all the devices in parallel
+                var visualComponents = _visualComponents.Values.ToArray();
+                var tasks = visualComponents.Select(v => v?.ViewModel.RefreshDeviceNameAsync());
```

Comment:
It isn't completely unnecessary, cause current implementation creates cross-layer loop: business logic -> visual component -> view model -> business logic. Also, I don't know if it matters from the performance perspective to have one request instead of many.

---
In reply to: [76335141](https://github.com/Microsoft/RTVS/pull/2199#discussion_r76335141) [](ancestors = 76335141)